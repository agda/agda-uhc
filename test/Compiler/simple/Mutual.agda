module Mutual where

open import Common.IO
open import Common.String
open import Common.Unit

mutual
  data G : Set where
    GA : {g : G}(f : F g) -> G
    GB : G

  data F : G -> Set where
    FA : (g : G) -> F g
    FB : F GB

mutual
  incG : G -> G
  incG GB     = GA FB
  incG (GA f) = GA (incF f)

  incF : {g : G} -> F g -> F (incG g)
  incF FB     = FA (GA FB)
  incF (FA g) = FA (incG g)




mutual
  PrintF : {g : G} -> F g -> String
  PrintF FB = "FB"
  PrintF (FA g) = "(FA " +S+ PrintG g +S+ ")"

  PrintG : G -> String
  PrintG GB     = "GB"
  PrintG (GA f) = "(GA " +S+ PrintF f +S+ ")"

main : IO Unit
main =
    putStrLn (PrintF (FA (GA (FA GB)))) ,,
    putStrLn (PrintG (incG (GA (incF FB)))) ,, --
    return unit
