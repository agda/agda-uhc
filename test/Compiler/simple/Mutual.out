EXECUTED_PROGRAM

ret > ExitSuccess
out > (FA (GA (FA GB)))
out > (GA (FA (GA (FA (GA FB)))))
out >
